### ISSUE TITLE
strings in the API

### ISSUE THREAD

**Issue by Xeverous** [cycfi/elements#54]
There is some discrepancy in the API how strings are taken.

https://github.com/cycfi/elements/blob/152401125dd77b229e40188bac872c4881b815fd/lib/include/elements/element/gallery/button.hpp#L147-L152

https://github.com/cycfi/elements/blob/152401125dd77b229e40188bac872c4881b815fd/lib/include/elements/element/gallery/check_box.hpp#L49

https://github.com/cycfi/elements/blob/152401125dd77b229e40188bac872c4881b815fd/lib/include/elements/element/gallery/dial.hpp#L25-L32

Is there any reason behind this? Having `std::string_view` or `std::string` (if widgets store strings internally) everywhere would improve consistency. I would like to not have `char const*` parameters.

**Comment by djowel** [comment]
> Is there any reason behind this? Having `std::string_view` or `std::string` (if widgets store strings internally) everywhere would improve consistency. I would like to not have `char const*` parameters.

If you look at the history, `std::string_view` was being added *incrementally* starting fairly recently (in terms of compiler years :-) ) with the switch to c++17. So I expect a lot of inconsistencies. I couldn't just change all APIs in one shot. 

Feel free to "fix" the inconsistencies.

**Comment by djowel** [comment]
> Feel free to "fix" the inconsistencies.

But please be careful. For example, `std::string_view` does not require the null terminator that some low level APIs (e.g. cairo) expects.

**Comment by Xeverous** [comment]
Ok. I will have some delay now, a disk in my computer died few days ago and currently using "spare/backup OS install" on a different drive. Downloaded all the tools but getting internal compiler error with building GCC :/

**Comment by Xeverous** [comment]
OK, so here is an idea following Sutter's CppCon talk "back to the basics: essentials of modern C++":

- All widgets which store strings should have a ctor/factory function taking `std::string` and moving from it.
  - This approach minimizes number of copies. The only case when a new string will be created is when the caller has no such object. But if we replace the ctor/factory function with `std::string_view` or `std::string const&`, the class will have to create a new string object always. With string value, an allocation can be avoided if the caller supplies the string object.
- All non-constructing widget functions should take `std::string_view`. **This includes `virtual void element::value(std::string_view)`** (a change affecting all widgets).
  - All such functions are already working on a widget with state. If the function took by a string by value, it would need to create a new string for every call (and destroy the internal state) unless the caller supplies one (which is usually a const lvalue reference). If the function takes a non-owning type (`std::string const&` or better with C++17: `std::string_view`) we avoid construction and reuse the existing in-class buffer.

So, in short:

- The ctor needs to create an object owned intenally. Take by value to allow passing rvalues and taking their ownership.
- Non-special member functions are working on an object with non-zero state. Take by const reference (or view such as `std::span`) because the internal state buffer can be reused. Callers of such functions rarely give rvalues.

Any thoughts? If we go for this, majority of widgets will need to have their ctor and `value()` rewritten.

**Comment by djowel** [comment]
That makes sense. Go for it!

**Comment by Xeverous** [comment]
Great. I will wait with this before #50 and #59 are merged to avoid conflicts.

**Comment by Xeverous** [comment]
Made #61. Intentionally avoided changing `std::string _font` and such as these might be rewritten at some point later.

**Comment by djowel** [comment]
Ah this one. Please add the rationale to #61 as well for reference.

**Comment by Xeverous** [comment]
PR merged, closing.
